### PR TITLE
Add a test package to the Idris 2 project

### PR DESCRIPTION
**Pull request by mattpolzin**
Closes https://github.com/idris-lang/Idris2/issues/1105.

This PR adds a `test` package to the Idris 2 project. It removes the recently duplicated `Golden` testing functionality from `tests/Lib.idr` and `libs/contrib/Test/Golden.idr` in favor of `libs/test/Test/Golden.idr`.

Removing the duplicated code removes opportunity for drift (for the record, the two files were already different as of this PR, though the differences were trivial to sort out).

Putting the resulting `Golden.idr` into a new testing package brings testing to the forefront and creates a home for modules that enable other styles of testing (unit and property-based, to name two non-mutually-exclusive examples).

There are changes in this PR to the way that Makefile steps depend on each other. Most prominently, building the test binary for the Idris 2 project is removed from the `all` step and `install` is added as a dependency of that test binary build step. This was done because the test binary is built and run against the version of Idris 2 installed on the system and at the same time building the test binary depends on libraries that might have been updated by the `make all` command.

The need to do this became apparent because the test binary cannot be built without the `test` package which is totally new as of this PR and therefore has not been installed before. Even without hitting a build failure like this, running tests without first installing newly built Idris files can result in misleading test failures (false negatives).

In our CI workflows, we always run `make install` before `make test` anyway, so in practice this has no impact on the guarantees made by CI. On the other hand, it eliminates the potential for someone running tests locally to mistakenly run `make all && make test` (with no install in between) which can result in failing tests cases that are false negatives.

### CODE REVIEW

File: package.nix [hunk 1]
```diff
@@ -31,7 +31,7 @@ stdenv.mkDerivation rec {
   # TODO: Move this into its own derivation, such that this can be changed
   #       without having to recompile idris2 every time.
   postInstall = let
-    includedLibs = [ "base" "contrib" "network" "prelude" ];
+    includedLibs = [ "base" "contrib" "network" "prelude" "testing" ];
```
Comment (zhongzc) [inline package.nix:34]: The difference between the module name `Test` and the package name `testing` increases the learning cost.

Comment (mattpolzin) [reply package.nix:34]: Fair point -- and after all, we have "network" and not "networking" as precedent for package naming.

File: tests/tests.ipkg [hunk 2]
```diff
@@ -1,5 +1,5 @@
 package runtests
 
-depends = contrib
+depends = contrib, test
```
Comment (gallais) [inline tests/tests.ipkg:3]: Do we still need the dependency on `contrib`?

Comment (mattpolzin) [reply tests/tests.ipkg:3]: Oh gosh, this comment is in tests. The answer must be no. Good catch.

**Comment by edwinb** [comment]
I realise I'm far too late on this one, but... I don't think `make test` should `install` first. I know that there's a problem with the libraries that `make test` depends on, but we should fix that instead. The problem with installing whenever you test is, if you run tests during development - which you should be doing regularly - and happen to have some problem that causes tests to fail, you've just installed a broken version which you're then stuck with (unless you've been careful in some other ways).

We need to sort out the test script to use the right libraries, certainly, but it shouldn't involve installing first.

**Comment by gallais** [comment]
Plus it's annoyingly slow to reinstall every time.

**Comment by mattpolzin** [comment]
Agreed; especially in hindsight, I solved the problem with testing against libraries different than the installed versions the wrong way around. If no one else has beat me to it, I’ll try to get this sorted out soon.

**Pull request merged**